client/src/components/R&R/NewReview.jsx

```diff
@@ -4,72 +4,279 @@ import $ from 'jquery';
-      count: 60,
+      count: `Minimum required characters left: 50`,
+      imgWindow: false,
+      submitted: false,
+      errors: {},
+      showError: false
     };
+
+    this.handleInputChange = this.handleInputChange.bind(this);
     this.toggleModal = this.toggleModal.bind(this);
+    this.toggleImgWindow = this.toggleImgWindow.bind(this);
     this.submitReview = this.submitReview.bind(this);
+    this.characterCounter = this.characterCounter.bind(this);
+    this.processCharSelection = this.processCharSelection.bind(this);
+  }
+
+  characterCounter(e) {
+    let charsLeft = 50 - e.target.value.length;
+    if (charsLeft > 0) {
+       this.setState({ count: `Minimum required characters left: ${charsLeft}` });
+    } else {
+      this.setState({ count: 'Minimum reached' });
+    }
+  }
+
+  handleInputChange(e, callback = () => {}) {
+    const name = e.target.getAttribute('name');
+    const value = e.target.value || e.target.getAttribute('value');
+
+    console.log('e.target >>>>', e.target);
+    this.setState({ [name]: value }, () => console.log(`${name} in state is now: `, this.state[name]), callback(e));
+  }
+
+  processCharSelection(name, value) {
+    console.log('this.props.characteristics >>>', this.props.characteristics);
+
+    const charId = this.props.characteristics[name].id
+    const update = {...this.state.characteristics, [charId]: Number(value) };
+    this.setState({characteristics: update}, () => console.log('state characteristics Obj >>>', this.state.characteristics));
+
   }
 
   submitReview() {
+    let errors = this.dataValidator();
+    console.log('errors >>>', errors);
+    if (Object.keys(errors).length > 0) {
+      this.setState({ errors, showError: true });
+    } else if (!this.state.submitted) {
+      this.setState({ showError: false });
+      const body = this.reviewConstructor();
+      handler.post(body, () => {
+        this.setState({
+          // show: false,
+          stars: null,
+          recommend: true,
+          characteristics: {},
+          summary: '',
+          body: '',
+          images: [],
+          nickname: '',
+          email: '',
+          count: `Minimum required characters left: 50`,
+          submitted: true,
+        });
+      });
+    }
+  }
+
+
+  reviewConstructor(data) {
+    const reqBody = {
+      product_id: this.props.pid,
+      rating: Number(this.state.stars),
+      summary: this.state.summary,
+      body: this.state.body,
+      recommend: Boolean(this.state.recommend),
+      name: this.state.nickname,
+      email: this.state.email,
+      characteristics: this.state.characteristics
+    }
+    console.log('reqbody >>>>', reqBody);
+    return reqBody;
+  }
+
+  dataValidator() {
+    console.log('current state >>>>', this.state);
+    let { stars, recommend, characteristics, summary, body, nickname, email, images } = this.state;
+    let errors = {};
 
+    if (stars === null) { errors.rating = ('Please select Overall Rating.'); }
```

Comment:
take a look at a library called Yup. it moves in the direction you solved for here, worth considering in future projects